src/analyses/mutexAnalysis.ml

```diff
@@ -36,17 +36,27 @@ struct
 
     module MakeG (G0: Lattice.S) =
     struct
-      module ReadWrite =
+      module ReadWriteNoRecover =
+      struct
+        include G0
+        let name () = "readwriteNoRecover"
+      end
+      module WriteNoRecover =
       struct
         include G0
-        let name () = "readwrite"
+        let name () = "writeNoRecover"
       end
-      module Write =
+      module ReadWriteRecover =
       struct
         include G0
-        let name () = "write"
+        let name () = "readwriteRecover"
       end
-      include Lattice.Prod (ReadWrite) (Write)
+      module WriteRecover =
+      struct
+        include G0
+        let name () = "writeRecover"
+      end
+      include Lattice.Prod4 (ReadWriteNoRecover) (WriteNoRecover) (ReadWriteRecover) (WriteRecover)
```

Comment:
Wouldn't something like
```ocaml
module P = Lattice.Prod (ReadWrite) (Write)
include Lattice.Prod (P) (P)
```
be easier because the recovery and non-recovery cases are symmetric?

Reply:
I am now trying an extensive-ish re-write also of check_fun, let's see if that is more clear then.

Reply:
I think it turned out quite nicely, let me know what you think!